### ISSUE TITLE
py3: proposal: at startup, check for correct UTF-8 setting. If not, do not start SABnzbd

### ISSUE THREAD

**Issue by sanderjo** [sabnzbd/sabnzbd#1336]
Based on https://github.com/sabnzbd/sabnzbd/issues/1335 I have this proposal:

Introduction: At SAB-py3 startup, SAB check if python version is at least 3.5. If not, SAB exits. See https://github.com/sabnzbd/sabnzbd/blob/develop/SABnzbd.py#L19-L22

My proposal: On POSIX, let's do such a hard check for ENV settings; check for UTF-8. If not, exit SABnzbd. Just like with an incorrect python version. Reason: Python3 seems to require UTF8 for correct behaviour, and I've discovered the traceback with non-UTF8 environment. See https://github.com/sabnzbd/sabnzbd/issues/1335 for the traceback.

I expect this check will only kick in on raw environments like docker images. 
And the solution is easy for the docker image creator:

`env LANG=C.UTF-8 LC_ALL=C.UTF-8 ./SABnzbd.py`
... where the locales must be available, which is the responsibility of OS / OS maintainer

It means the check https://github.com/sabnzbd/sabnzbd/blob/develop/SABnzbd.py#L1115-L1117 (so `locale.getpreferredencoding()` ) more to the front, and exit if not correct (not just a warning, like now).

Even my old Synology has a correct setting:
```
admin@DiskStation:/$ python3 -c "import locale; print(locale.getpreferredencoding())"
UTF-8
admin@DiskStation:/$ locale | grep -e "LANG=" -e "LC_ALL"
LANG=en_US.utf8
LC_ALL=en_US.utf8
```

**Comment by sanderjo** [comment]
Code:

```
#!/usr/bin/env python3

import os
import locale
import sys

'''
Bad:
posix
ANSI_X3.4-1968
'''

#print(os.name)
#print(locale.getpreferredencoding())

if ('posix' in os.name) and not ('utf-8' in locale.getpreferredencoding().lower()):
    print("No UTF-8 environment. Check your LANG and LC_ALL settings. Exiting.")
    sys.exit(1)

print("All good ...")
```
Results:
```

root@de8831c39769:/# ./checkit.py 
No UTF-8 environment. Check your LANG and LC_ALL settings. Exiting.
```
resp
```
root@de8831c39769:/# env LANG=C.UTF-8 LC_ALL=C.UTF-8 ./checkit.py 
All good ...
```

**Comment by Safihre** [comment]
Sounds good. Let's use the same error text that we had before in the py2 version. It's in the `2.3.x` branch on Github still. In `SABnzbd.py` startup bit. I removed it in py3 version because I thought it wasn't needed.

**Comment by sanderjo** [comment]
OK.

But which "error text" do you mean? **Not** the "SABnzbd was started with encoding %s, this should be UTF-8. Expect problems with Unicoded file and directory names in downloads.", I presume?

**Comment by Safihre** [comment]
Yes that one. That's wrong?

**Comment by sanderjo** [comment]
Result with altered SABnzbd:

```
root@de8831c39769:/# sabnzbdplus
No UTF-8 environment. Check your LANG and LC_ALL settings. Exiting.
```

resp
```
root@de8831c39769:/# env LANG=C.UTF-8 LC_ALL=C.UTF-8 sabnzbdplus
2019-10-16 18:34:39,910::INFO::[sabnzbdplus:1105] --------------------------------
2019-10-16 18:34:39,911::INFO::[sabnzbdplus:1106] sabnzbdplus-3.0.0-develop (rev=unknown)
2019-10-16 18:34:39,911::INFO::[sabnzbdplus:1107] Full executable path = /usr/bin/sabnzbdplus
2019-10-16 18:34:39,911::INFO::[sabnzbdplus:1117] Platform = posix
2019-10-16 18:34:39,912::INFO::[sabnzbdplus:1118] Python-version = 3.6.8 (default, Oct  7 2019, 12:59:55) 
[GCC 8.3.0]
2019-10-16 18:34:39,912::INFO::[sabnzbdplus:1119] Arguments = /usr/bin/sabnzbdplus
2019-10-16 18:34:39,912::INFO::[sabnzbdplus:1122] Preferred encoding = UTF-8
2019-10-16 18:34:39,912::INFO::[sabnzbdplus:1129] SSL version = OpenSSL 1.1.1  11 Sep 2018
2019-10-16 18:34:39,912::INFO::[sabnzbdplus:1176] Using INI file /root/.sabnzbd/sabnzbd.ini
```

**Comment by Safihre** [comment]
So SABnzbd can't be loaded from the web when started non-UTF8?
Hmmm this makes it difficult. In the NAS-cases for example, how would the user be notified of a problem..

**Comment by thezoggy** [comment]
Majority of people just install via binary or some sort of package provider.
This scenario should be rare in this day in age, and back in sb dev time (where sab got the current code from for handling this) we only ran into it when a user was trying to setup linux and didn't really know what they were doing. Fast forward to now, if their env is broken other stuff is probably broken. We should abort and log/print the error and not try to continue in a broken state.

To prove the point, how many times have we had people complain about sab not working because of this today....  which would have a higher chance of an state with string/unicode problems in py2.

**Comment by sanderjo** [comment]
I agree with @thezoggy

**Comment by Safihre** [comment]
Oke! 👍Could you provide a pull request with your first code to exit straight away?

**Comment by sanderjo** [comment]
> Yes that one. That's wrong?

That text is too friendly; not enough to explain the exit.

**Comment by sanderjo** [comment]
> Oke! Could you provide a pull request with your first code to exit straight away?

Yes, I'll do. Hopefully it will not interfere with my ongoing work on the de-obfuscating code.

**Comment by sanderjo** [comment]
Weird ... after 1 hour of trying:
- I **cannot** reproduce the traceback with the git-py3 version if the ENV settings are non-UTF
- I can reproduce the traceback with @jcfp 's ppa py3 version.

So ... how can that be? Difference within sabnzbd itself ... or in the libraries/requirements?

**Comment by sanderjo** [comment]
Because this is so unbelievable, I've main a screencast:

https://youtu.be/_Z7bL8bQaSI

**Comment by jcfp** [comment]
@sanderjo as noted in [#1335](https://github.com/sabnzbd/sabnzbd/issues/1335#issuecomment-540168302), the traceback appears to be introduced by changes made as part of packaging. Just haven't gotten around to figuring out what exactly.

**Comment by sanderjo** [comment]
> @sanderjo as noted in [#1335](https://github.com/sabnzbd/sabnzbd/issues/1335#issuecomment-540168302), the traceback appears to be introduced by changes made as part of packaging.

Hmmm ... so I've missed a very (the most?) important remark by you ... Sorry.

**Comment by sanderjo** [comment]
Because of the error message
`UnicodeDecodeError: 'ascii' codec can't decode byte 0xef in position 8593: ordinal not in range(128)
`
I tried to find 0xef in ASCII files in the Glitter directory, but could not find it.


So, then, in /usr/share/sabnzbdplus/interfaces (so the PPA stuff):
- I moved away the Glitter directory
- I copied in the Glitter from the sab-git source

... and it's working now: no more traceback with non-UTF ENV with sanbzd-ppa version.

So ... this is indeed something in the PPA version, not in the mainstream git version. So I'll close this  (non) issue.

**Comment by jcfp** [comment]
@sanderjo at least resurrecting the issue made me finally work on it :grin:

The cause is the minified version of jquery.js from debian pkg libjs-jquery/3.2.1-1, which in the packaged sab version replaces the minified file included with sab releases (taken from 2.2.4 jquery upstream). Pointing the symlink at /usr/share/sabnzbdplus/interfaces/Glitter/templates/static/javascripts/jquery.min.js to the non-minified version instead makes the traceback go away.

Trigger seemed to be [this piece of code](https://github.com/jquery/jquery/blob/30e1bfbdcb0ff658f1fa128b72480194e8ecb926/src/selector/escapeSelector.js#L16), something goes awry when that \uFFFD is minified, or cherrypy just can't handle that...

Either way, this should be an easy fix in the packaging.

**Comment by sanderjo** [comment]
On https://launchpad.net/~jcfp/+archive/ubuntu/private/+packages I see

```
Changelog
sabnzbdplus (3.0.0~alpha0~git20191001.1300-0ubuntu1~jcfp5) bionic; urgency=medium

  * Links: point Glitter jquery.min.js link to the non-minified
    javascript file provided by libjs-jquery. For background, see
    https://github.com/sabnzbd/sabnzbd/issues/1336#issuecomment-554095620

 -- JCF Ploemen (jcfp) <email address hidden>  Thu, 14 Nov 2019 21:49:22 +0000
```

and indeed, after a apt-get update, sabnzbdplus has no Traceback when starting with

`env LC_ALL=C LANG=C sabnzbdplus
`

So, @jcfp , thank you for solving this! And thank you for providing this PPA ... it makes using and thus testing SAB-py3 so much easier!
